xswap/hash_table.cpp

```diff
@@ -0,0 +1,46 @@
+#include <cstdlib>
+# include <cstring>
+#include "hash_table.h"
+
+EdgeHashTable::EdgeHashTable(int max_source, int max_target) {
```

Comment:
A few general comments on this class:
- Can you add some comments on `get_bit`, `set_bit_true` and `set_bit_false` methods? By reading the code, I don't understand why the bit is get/set at `7 - bit_position`.
- You may want to add a destructor to call `free()` to avoid memory leak of `hash_table`.
- Have you tried to implement this class with `bitset` in C++ standard library?  http://www.cplusplus.com/reference/bitset/bitset/set/
It has a friendly API and you don't need to worry about memory allocation. I don't know whether it will hurt the performance though. (If you don't have time to try it, please ignore this comment.)

Reply:
By the way, `std::vector<bool>` is another alternative, which is implemented with bit instead of boolean type to save space.

Reply:
I will attempt to implement `bitset` in a future PR, once I have some tests written than can check its methods are equivalent and similarly fast.